src/physics/em/detail/RayleighInteractor.i.hh

```diff
@@ -15,54 +18,119 @@ namespace detail
  * Construct with shared and state data.
  */
 CELER_FUNCTION
-RayleighInteractor::RayleighInteractor(const RayleighInteractorPointers& shared,
-                                       const ParticleTrackView& particle,
-                                       const Real3&             inc_direction,
-                                       StackAllocator<Secondary>& allocate)
+RayleighInteractor::RayleighInteractor(const RayleighNativePointers& shared,
+                                       const ParticleTrackView&      particle,
+                                       const Real3&       inc_direction,
+                                       const ElementView& element)
     : shared_(shared)
     , inc_energy_(particle.energy().value())
     , inc_direction_(inc_direction)
-    , allocate_(allocate)
+    , element_(element)
 {
     CELER_EXPECT(inc_energy_ >= this->min_incident_energy()
                  && inc_energy_ <= this->max_incident_energy());
-    CELER_EXPECT(particle.particle_id() == shared_.gamma_id); // XXX
-    CELER_NOT_IMPLEMENTED("Rayleigh scattering");
+    CELER_EXPECT(particle.particle_id() == shared_.gamma_id);
 }
 
 //---------------------------------------------------------------------------//
 /*!
- * Sample using the XXX model.
+ * Sample using the G4LivermoreRayleighModel model.
  */
 template<class Engine>
 CELER_FUNCTION Interaction RayleighInteractor::operator()(Engine& rng)
 {
-    // Allocate space for XXX (electron, multiple particles, ...)
-    Secondary* secondaries = this->allocate_(0); // XXX
-    if (secondaries == nullptr)
-    {
-        // Failed to allocate space for a secondary
-        return Interaction::from_failure();
-    }
-
-    // XXX sample
-    (void)sizeof(rng);
+    real_type energy = inc_energy_.value();
 
     // Construct interaction for change to primary (incident) particle
     Interaction result;
-    result.action      = Action::scattered;                     // XXX
-    result.energy      = units::MevEnergy{inc_energy_.value()}; // XXX
-    result.direction   = inc_direction_;
-    result.secondaries = {secondaries, 1}; // XXX
+    result.action = Action::scattered;
+    result.energy = units::MevEnergy{energy};
+
+    // Sample direction for a given Z: G4RayleighAngularGenerator
+    unsigned int Z = element_.atomic_number();
+    using ItemIdT  = celeritas::ItemId<int>;
+    Array<real_type, rayleigh_num_parameters> params
+        = shared_.params.data[ItemIdT{Z - 1}];
```

Comment:
For our existing datasets, rather than copying all elements to the GPU and indexing into them with `Z - 1`, we've looped over `MaterialParams::num_elements()` and copied over only the elements in use in the current problem, and indexed into collections with `element_id`.

In practice I don't think it'll make too much of a difference vs copying over everything, but it's definitely safer to access the data via `element_id` rather than an offset Z integer.

Reply:
Add MaterialParams in RayleighModel and store angular parameters only for need element with element_id.